pkg/apis/core/types.go

```diff
@@ -2391,6 +2392,13 @@ type PodAffinityTerm struct {
 	// selected pods is running.
 	// Empty topologyKey is not allowed.
 	TopologyKey string
+	// Selects Namespaces using cluster scoped-labels.  This
+	// matches namespaces selected by this label selector.
+	// This field follows standard label selector semantics.
+	// If present but empty, this selector selects all namespaces,
```

Comment:
were the scale issues with enforcing affinity/anti-affinity rules across all namespaces resolved? cc @wojtek-t @shyamjvs

Reply:
Not sure I understand - this PR seems to be the one adding them for namespaces, right? If you meant scale issues seen with pod anti-affinity across 'nodes', it was still not working satisfactorily on large scale the last I remember (ref https://github.com/kubernetes/kubernetes/issues/54164).

However, I guess @bsalamat who was working on optimizing it is more up-to-date here than me.

Reply:
affinity/anti-affinity still has room for performance improvement, but this PR is orthogonal to that effort.

Reply:
> this PR is orthogonal to that effort.

it seems directly related... this adds another loop that looks at every namespace around pod affinity evaluation, and makes it trivial to select every namespace in the cluster

Reply:
Jordan, you are right that someone could create rules that makes performance even worse than before, for example, by choosing all namespaces, but even without this new field, our implementation has to go over all pods in all namespaces and check whether they match any of the "Namespaces" and "LablelSelectors". So, I believe this feature won't cause a significant slow down, unless I am missing something.

Reply:
> even without this new field, our implementation has to go over all pods in all namespaces and check whether they match any of the "Namespaces" and "LablelSelectors". So, I believe this feature won't cause a significant slow down, unless I am missing something.

I agree the situation is pretty bad already, but this adds *another* O(n) evaluation. Previously, it was O(1) to get the namespaces to evaluate for a pod. Now it requires evaluating a label selector against every namespace that exists.

Reply:
It is not another O(n) per pod. It is another O(n) per affinity term (where n is the number of namespaces) and once we find the namespaces, the rest of the logic and runtime complexity remains the same. That is we go over all the pods and check if they match any of the namespaces. Assuming that n is much smaller than the number of Pods in a cluster, I think this is not a significant slow down.

Reply:
> It is not another O(n) per pod. It is another O(n) per affinity term 

Affinity term is a per-pod field 

> Assuming that n is much smaller than the number of Pods in a cluster

That's not a valid assumption. 

The discussion around the affinity feature performance seemed to agree that there were existing and significant performance/scalability concerns. Is there agreement from the scheduling/scalability teams to continue increasing the complexity and scope of affinity evaluations before the existing feature set has been shown to scale?

Reply:
> Affinity term is a per-pod field

To be more precise, I am talking about [this function](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/predicates.go#L1160). We first get the namespaces and then loop over all pods and check, but as I said from the beginning, you are right that this adds to the complexity.

> That's not a valid assumption.

At least nn order of magnitude is pretty common in most clusters.

> The discussion around the affinity feature performance seemed to agree that there were existing and significant performance/scalability concerns. Is there agreement from the scheduling/scalability teams to continue increasing the complexity and scope of affinity evaluations before the existing feature set has been shown to scale?

We are definitely not interested in increasing the complexity. This change is rather a convenience feature in the API. Even without this change, one may list all namespaces in the existing "namespaces" field of the affinity term. This feature tries to make the API more consistent with the rest of K8s API. That said, I don't disagree that it may be better to address the performance issue of the affinity feature before making any more changes.